src/codegen/general.rs

```diff
@@ -16,32 +16,71 @@ use std::{
 
 pub fn start_comments(w: &mut dyn Write, conf: &Config) -> Result<()> {
     if conf.single_version_file.is_some() {
-        return start_comments_no_version(w);
+        return start_comments_no_version(w, conf);
     }
-    writeln!(
+    write!(
         w,
         "// This file was generated by gir (https://github.com/gtk-rs/gir @ {})
-// from gir-files (https://github.com/gtk-rs/gir-files @ {})
-// DO NOT EDIT",
-        VERSION, conf.girs_version
+{}// DO NOT EDIT",
+        VERSION,
+        conf.girs_version
+            .iter()
+            .map(|info| {
+                format!(
+                    "// from {} ({}{})\n",
+                    info.gir_dir.display(),
+                    match info.get_repository_url() {
+                        Some(url) => format!("{} @ ", url),
+                        None => String::new(),
+                    },
+                    info.get_hash(),
+                )
+            })
+            .collect::<String>(),
     )
 }
 
-pub fn start_comments_no_version(w: &mut dyn Write) -> Result<()> {
+pub fn start_comments_no_version(w: &mut dyn Write, conf: &Config) -> Result<()> {
     writeln!(
         w,
         "// This file was generated by gir (https://github.com/gtk-rs/gir)
-// from gir-files (https://github.com/gtk-rs/gir-files)
-// DO NOT EDIT"
+{}// DO NOT EDIT",
+        conf.girs_version
+            .iter()
+            .map(|info| {
+                format!(
+                    "// from {}{}\n",
+                    info.gir_dir.display(),
+                    match info.get_repository_url() {
+                        Some(url) => format!(" ({})", url),
+                        None => String::new(),
+                    },
+                )
+            })
+            .collect::<String>()
     )
 }
 
 pub fn single_version_file(w: &mut dyn Write, conf: &Config) -> Result<()> {
-    writeln!(
+    write!(
         w,
         "Generated by gir (https://github.com/gtk-rs/gir @ {})
-from gir-files (https://github.com/gtk-rs/gir-files @ {})",
-        VERSION, conf.girs_version
+{}",
+        VERSION,
+        conf.girs_version
+            .iter()
+            .map(|info| {
+                format!(
+                    "from {} ({}@ {})\n",
+                    info.gir_dir.display(),
+                    match info.get_repository_url() {
+                        Some(url) => format!("{} @ ", url),
```

Comment:
Here you're emitting double `@`'s. Just move the code from `start_comments` into `single_version_file` and call that instead.

Reply:
@GuillaumeGomez Whoops, forgot to say that `single_version_file` does not emit `//` before every line, breaking all rust files :hand_over_mouth:

Reply:
The easy fix is of course:

```diff
diff --git a/src/codegen/general.rs b/src/codegen/general.rs
index e956c4f..de75ddd 100644
--- a/src/codegen/general.rs
+++ b/src/codegen/general.rs
@@ -18,7 +18,7 @@ pub fn start_comments(w: &mut dyn Write, conf: &Config) -> Result<()> {
     if conf.single_version_file.is_some() {
         start_comments_no_version(w, conf)
     } else {
-        single_version_file(w, conf)
+        single_version_file(w, conf, "// ")
     }
 }

@@ -41,17 +41,19 @@ pub fn start_comments_no_version(w: &mut dyn Write, conf: &Config) -> Result<()>
     )
 }

-pub fn single_version_file(w: &mut dyn Write, conf: &Config) -> Result<()> {
+pub fn single_version_file(w: &mut dyn Write, conf: &Config, prefix: &str) -> Result<()> {
     write!(
         w,
-        "Generated by gir (https://github.com/gtk-rs/gir @ {})
+        "{}Generated by gir (https://github.com/gtk-rs/gir @ {})
 {}",
+        prefix,
         VERSION,
         conf.girs_version
             .iter()
             .map(|info| {
                 format!(
-                    "from {} ({}@ {})\n",
+                    "{}from {} ({}@ {})\n",
+                    prefix,
                     info.gir_dir.display(),
                     info.get_repository_url()
                         .map_or_else(String::new, |u| format!("{} ", u)),
diff --git a/src/codegen/mod.rs b/src/codegen/mod.rs
index 6303210..177e3be 100644
--- a/src/codegen/mod.rs
+++ b/src/codegen/mod.rs
@@ -72,7 +72,7 @@ pub fn generate_mod_rs(env: &Env, root_path: &Path, mod_rs: &[String], traits: &
 pub fn generate_single_version_file(env: &Env) {
     if let Some(ref path) = env.config.single_version_file {
         save_to_file(path, env.config.make_backup, |w| {
-            general::single_version_file(w, &env.config)
+            general::single_version_file(w, &env.config, "")
         });
     }
 }
```

Reply:
We lost `// DO NOT EDIT` as well, btw.